engine/common/provider/engine.go

```diff
@@ -224,3 +250,66 @@ func (e *Engine) onEntityRequest(originID flow.Identifier, req *messages.EntityR
 
 	return nil
 }
+
+func (e *Engine) processQueuedRequestsShovellerWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	e.log.Debug().Msg("process entity request shoveller worker started")
+
+	for {
+		select {
+		case <-e.requestHandler.GetNotifier():
+			// there is at least a single request in the queue, so we try to process it.
+			e.shovelEntityRequests()
+		case <-ctx.Done():
+			// close the internal channel, the workers will drain the channel before exiting
+			close(e.requestChannel)
+			e.log.Trace().Msg("processing entity request worker terminated")
+			return
+		}
+	}
+}
+
+func (e *Engine) shovelEntityRequests() {
+	for {
+		msg, ok := e.requestQueue.Get()
+		if !ok {
+			// no more requests, return
+			return
+		}
+
+		req := &internal.EntityRequest{
+			OriginId:  msg.OriginID,
+			EntityIds: msg.Payload.([]flow.Identifier),
+		}
+
+		lg := e.log.With().
+			Hex("origin_id", logging.ID(req.OriginId)).
+			Str("requested_entity_ids", fmt.Sprintf("%v", req.EntityIds)).Logger()
+
+		lg.Trace().Msg("shoveller is queuing entity request for processing")
+		e.requestChannel <- req
+		lg.Trace().Msg("shoveller queued up entity request for processing")
+	}
+}
+
+func (e *Engine) processEntityRequestWorker(_ irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	for {
+		request, ok := <-e.requestChannel
+		if !ok {
+			e.log.Trace().Msg("processing entity request worker terminated")
+			return
+		}
+		lg := e.log.With().
+			Hex("origin_id", logging.ID(request.OriginId)).
+			Str("requested_entity_ids", fmt.Sprintf("%v", request.EntityIds)).Logger()
+		lg.Trace().Msg("worker picked up entity request for processing")
+		err := e.onEntityRequest(request.OriginId, request.EntityIds)
```

Comment:
Please handle expected errors here, and `ctx.Throw` unexpected errors.

Reply:
https://github.com/onflow/flow-go/pull/3404/commits/e54fc258a0f093fe6a15308dec7ae30d7c252910